File: RayTracer/Camera.fs
module RayTracer.Camera

open System
open System.Numerics
open Ray
open Color

type Camera =
    { origin: Vector3 // x, y, z
      direction: Quaternion // yaw, pitch, roll
      viewAngle: Vector3 } // horizontal, vertical, distance

module Camera =
    let init (origin: Vector3) (direction: Quaternion) (viewAngle: Vector3) =
        { origin = origin
          direction = direction
          viewAngle = viewAngle }

    let moveCamera (camera: Camera) (origin: Vector3): Camera = { camera with origin = origin }

    let castRay (camera: Camera) (pixel: Vector2): Ray =
        let direction =
            Vector3(pixel.X * camera.viewAngle.X, pixel.Y * camera.viewAngle.Y, camera.viewAngle.Z)
            |> fun d -> Vector3.Transform(d, camera.direction)
            |> Vector3.Normalize

        { origin = camera.origin
          direction = direction
          color = SpectralColor.initOne }

File: RayTracerCs/Camera.cs
using System.Numerics;

namespace RayTracerCs
{
    public class Camera
    {
        public Camera(Vector3 origin, Quaternion direction, Vector3 viewAngle)
        {
            Origin = origin;
            Direction = direction;
            ViewAngle = viewAngle;
        }

        public Vector3 Origin { get; }
        public Quaternion Direction { get; }
        public Vector3 ViewAngle { get; }

        public Ray CastRay(Vector2 pixel)
        {
            return new Ray(
                Origin,
                Vector3.Normalize(
                    Vector3.Transform(
                        new Vector3(pixel.X * ViewAngle.X, pixel.Y * ViewAngle.Y, ViewAngle.Z),
                        Direction
                    )
                ),
                SpectralColor.One()
            );
        }
    }
}